File: raytrace/Parse/PlyLoader.cpp
#include "Parse/PlyLoader.hpp"

#include "Object/Shape/TriangleMesh.hpp"

#include "Math/Point.hpp"

#include <fstream>
#include <string>

namespace Parse {
	struct Property {
		std::string name;
		std::string type;
		std::string listType;
	};
	struct Element {
		std::string name;
		int count;
		std::vector<Property> properties;
	};

	std::vector<Element> parseElements(std::ifstream &file)
	{
		std::vector<Element> elements;
		Element newElement;
		std::string keyword;
		file >> keyword;

		while (keyword != "end_header") {
			if (keyword == "comment") {
				std::string line;
				std::getline(file, line);
				file >> keyword;
				continue;
			}
			else if (keyword == "element") {
				file >> newElement.name >> newElement.count;

				file >> keyword;
				while (keyword == "property") {
					Property property;
					file >> property.type;
					if (property.type == "list") {
						std::string indexType;
						file >> indexType >> property.listType;
					}
					file >> property.name;
					newElement.properties.push_back(std::move(property));
					file >> keyword;
				}
				elements.push_back(std::move(newElement));
			}
		}

		return elements;
	}

	void skipProperty(std::ifstream &file, const Property &property) {
		std::string text;
		if (property.type == "list") {
			int count;
			file >> count;
			for (int i = 0; i < count; i++) {
				file >> text;
			}
		}
		else {
			file >> text;
		}
	}

	void skipElement(std::ifstream &file, const Element &element) {
		for (int i = 0; i < element.count; i++) {
			for (const Property &property : element.properties) {
				skipProperty(file, property);
			}
		}
	}

	std::unique_ptr<Object::Shape::Base> PlyLoader::load(const std::string &filename)
	{
		std::ifstream file(filename.c_str());
		std::string ply;

		file >> ply;
		if (ply != "ply") {
			return std::unique_ptr<Object::Shape::Base>();
		}

		std::string format, type, version;
		file >> format >> type >> version;
		if (format != "format" || type != "ascii") {
			return std::unique_ptr<Object::Shape::Base>();
		}

		std::vector<Object::Shape::TriangleMesh::Vertex> vertices;
		std::vector<Object::Shape::TriangleMesh::Triangle> triangles;

		std::vector<Element> elements = parseElements(file);
		for (const Element &element : elements) {
			if (element.name == "vertex") {
				int idxX = -1;
				int idxY = -1;
				int idxZ = -1;
				for (int i = 0; i < element.properties.size(); i++) {
					if (element.properties[i].name == "x") {
						idxX = i;
					}
					else if (element.properties[i].name == "y") {
						idxY = i;
					}
					else if (element.properties[i].name == "z") {
						idxZ = i;
					}
				}

				for (int i = 0; i < element.count; i++) {
					float x = 0;
					float y = 0;
					float z = 0;
					for (int j = 0; j < element.properties.size(); j++) {
						if (j == idxX) {
							file >> x;
						}
						else if (j == idxY) {
							file >> y;
						}
						else if (j == idxZ) {
							file >> z;
						}
						else {
							skipProperty(file, element.properties[j]);
						}
					}

					Object::Shape::TriangleMesh::Vertex vertex;
					vertex.point = Math::Point(x, y, z);
					vertices.push_back(vertex);
				}
			}
			else if (element.name == "face") {
				int idxVertices = -1;
				for (int i = 0; i < element.properties.size(); i++) {
					if (element.properties[i].name == "vertex_indices") {
						idxVertices = i;
					}
				}

				for (int i = 0; i < element.count; i++) {
					Object::Shape::TriangleMesh::Triangle triangle;
					for (int j = 0; j < element.properties.size(); j++) {
						if (j == idxVertices) {
							int count;
							file >> count;
							for (int k = 0; k < count; k++) {
								if (k < 3) {
									file >> triangle.vertices[k];
								}
								else {
									Math::Vector u = vertices[triangle.vertices[1]].point - vertices[triangle.vertices[0]].point;
									Math::Vector v = vertices[triangle.vertices[2]].point - vertices[triangle.vertices[0]].point;
									triangle.normal = Math::Normal(u % v).normalize();
									triangles.push_back(triangle);
									triangle.vertices[1] = triangle.vertices[2];
									file >> triangle.vertices[2];
								}
							}
							Math::Vector u = vertices[triangle.vertices[1]].point - vertices[triangle.vertices[0]].point;
							Math::Vector v = vertices[triangle.vertices[2]].point - vertices[triangle.vertices[0]].point;
							triangle.normal = Math::Normal(u % v).normalize();
							triangles.push_back(triangle);
						}
						else {
							skipProperty(file, element.properties[j]);
						}
					}
				}
			}
			else {
				skipElement(file, element);
			}
		}

		return std::make_unique<Object::Shape::TriangleMesh>(std::move(vertices), std::move(triangles));
	}
}

File: src/input/ply_loader.rs
use crate::geo;
use geo::Normal3;
use geo::Point3;

use crate::object;
use object::Shape;

use object::shape::TriangleMesh;
use object::shape::TriangleMeshVertex;
use object::shape::TriangleMeshTriangle;

use std::io::BufReader;
use std::io::BufRead;
use std::fs::File;

use std::collections::VecDeque;

pub struct PlyLoader;

struct ItemReader {
    reader: BufReader<File>,
    items: VecDeque<String>
}

impl ItemReader {
    pub fn new(reader: BufReader<File>) -> ItemReader {
        return ItemReader { reader, items: VecDeque::new() };
    }

    pub fn next(&mut self) -> String {
        if self.items.len() == 0 {
            let mut line = String::new();
            if let Ok(_size) = self.reader.read_line(&mut line) {
                for item in line.split_whitespace() {
                    self.items.push_back(item.to_string());
                }
            }
        }

        return self.items.pop_front().unwrap_or_default();
    }

    pub fn expect_item(&mut self, target: &str) -> Option<()> {
        let item = self.next();
        if item == target {
            return Some(());
        } else {
            return None;
        } 
    }

    pub fn eat_line(&mut self) {
        self.items.clear();
    }
}

struct Property {
    pub name: String,
    pub property_type: String,
    pub _list_type: String
}
struct Element {
    pub name: String,
    pub count: usize,
    pub properties: Vec<Property>
}

impl PlyLoader {
    pub fn load(filename: &str) -> Option<Box<dyn Shape>> {
        let f = File::open(filename);
        if let Err(_) = f {
            return None;
        }
        let f = f.unwrap();
        let b = BufReader::new(f);
        let mut reader = ItemReader::new(b);

        reader.expect_item("ply")?;
        reader.expect_item("format")?;
        reader.expect_item("ascii")?;
        let _ = reader.next();

        let elements = Self::parse_elements(&mut reader);
        let mut vertices = Vec::new();
        let mut triangles = Vec::new();

        for element in &elements {
            if element.name == "vertex" {
                let mut idx_x = 0;
                let mut idx_y = 0;
                let mut idx_z = 0;
                for (idx, property) in element.properties.iter().enumerate() {
                    match property.name.as_str() {
                        "x" => idx_x = idx,
                        "y" => idx_y = idx,
                        "z" => idx_z = idx,
                        _ => ()
                    }
                }

                for _ in 0..element.count {
                    let mut x = 0.0;
                    let mut y = 0.0;
                    let mut z = 0.0;
                    for j in 0..element.properties.len() {
                        if j == idx_x {
                            x = reader.next().parse::<f32>().unwrap_or_default();
                        } else if j == idx_y {
                            y = reader.next().parse::<f32>().unwrap_or_default();
                        } else if j == idx_z {
                            z = reader.next().parse::<f32>().unwrap_or_default();
                        } else {
                            Self::skip_property(&mut reader, &element.properties[j]);
                        }
                    }

                    let vertex = TriangleMeshVertex { point: Point3::new(x, y, z) };
                    vertices.push(vertex);
                }
            } else if element.name == "face" {
                let mut idx_vertices = 0;
                for (idx, property) in element.properties.iter().enumerate() {
                    if property.name == "vertex_indices" {
                        idx_vertices = idx;
                    }
                }

                for _ in 0..element.count {
                    for j in 0..element.properties.len() {
                        if j == idx_vertices {
                            let mut triangle_vertices = [0 as usize; 3];
                            let count = reader.next().parse::<usize>().unwrap_or_default();
                            for k in 0..count {
                                let index = reader.next().parse::<usize>().unwrap_or_default();
                                if k < 3 {
                                    triangle_vertices[k] = index;
                                } else {
                                    let u = vertices[triangle_vertices[1]].point - vertices[triangle_vertices[0]].point;
                                    let v = vertices[triangle_vertices[2]].point - vertices[triangle_vertices[0]].point;
                                    let normal = Normal3::from(u % v).normalize();
                                    let triangle = TriangleMeshTriangle { vertices: [triangle_vertices[0], triangle_vertices[1], triangle_vertices[2]], normal };
                                    triangles.push(triangle);
                                    triangle_vertices[1] = triangle_vertices[2];
                                    triangle_vertices[2] = index;
                                }
                            }

                            let u = vertices[triangle_vertices[1]].point - vertices[triangle_vertices[0]].point;
                            let v = vertices[triangle_vertices[2]].point - vertices[triangle_vertices[0]].point;
                            let normal = Normal3::from(u % v).normalize();
                            let triangle = TriangleMeshTriangle { vertices: [triangle_vertices[0], triangle_vertices[1], triangle_vertices[2]], normal };
                            triangles.push(triangle);
                        } else {
                            Self::skip_property(&mut reader, &element.properties[j]);
                        }
                    }
                }
            } else {
                Self::skip_element(&mut reader, &element);
            }
        }

        return Some(Box::new(TriangleMesh::new(vertices, triangles)));
    }

    fn parse_elements(reader: &mut ItemReader) -> Vec<Element> {
        let mut elements = Vec::new();
        let mut keyword = reader.next();
        while keyword != "end_header" {
            if keyword == "comment" {
                reader.eat_line();
                keyword = reader.next();
            } else if keyword == "element" {
                let name = reader.next();
                let count = reader.next().parse::<usize>().unwrap_or_default();
                let mut properties = Vec::new();
                keyword = reader.next();
                while keyword == "property" {
                    let property_type = reader.next();
                    let mut _list_type = String::new();
                    if property_type == "list" {
                        let _index_type = reader.next();
                        _list_type = reader.next();
                    }
                    let name = reader.next();
                    let property = Property { name, property_type, _list_type };
                    properties.push(property);
                    keyword = reader.next();
                }
                let element = Element { name, count, properties };
                elements.push(element);
            }
        }

        return elements;
    }

    fn skip_property(reader: &mut ItemReader, property: &Property) {
        if property.property_type == "list" {
            let count = reader.next().parse::<usize>().unwrap_or_default();
            for _ in 0..count {
                let _item = reader.next();
            }
        } else {
            let _item = reader.next();
        }
    }
    
    fn skip_element(reader: &mut ItemReader, element: &Element) {
        for _ in 0..element.count {
            for property in &element.properties {
                Self::skip_property(reader, property);
            }
        }
    }
}